internal/hrq/hrqreconciler.go

```diff
@@ -156,6 +156,24 @@ func isDeleted(inst *api.HierarchicalResourceQuota) bool {
 	return inst.GetCreationTimestamp() == (metav1.Time{})
 }
 
+// OnChangeNamespace enqueues HRQ objects for later reconciliation. This occurs in a
+// goroutine so the caller doesn't block; since the reconciler is never garbage-collected, this is safe.
+func (r *HierarchicalResourceQuotaReconciler) OnChangeNamespace(log logr.Logger, ns *forest.Namespace) {
+	go func() {
+		r.Forest.Lock()
+		defer r.Forest.Unlock()
+
+		inst := &api.HierarchicalResourceQuota{}
+		for _, nsnm := range ns.AncestryNames() {
+			for _, hrqnm := range r.Forest.Get(nsnm).HRQNames() {
+				inst.ObjectMeta.Name = hrqnm
+				inst.ObjectMeta.Namespace = nsnm
+				r.trigger <- event.GenericEvent{Object: inst}
```

Comment:
This line can block so we shouldn't be holding the forest lock by the time we call it (that's why this method's in a goroutine in the first place). However, the forest lock should be called when `OnChangeNamespace` is called in the first place. So you could create a slice of all the HRQs before starting the goroutine (which means the lock is held), and then just iterate over that slice in the goroutine.

Reply:
Understood. Will do.